packages/stack/proxy/src/rpc-manager.ts

```diff
@@ -0,0 +1,193 @@
+
+  public _accounts: Account[] | null = null;
+
+  public _nodeAccounts: string[] | null = null;
+
+  private requestInterceptors: Array<RpcRequestRegistration<any>> = [];
+
+  private responseInterceptors: Array<RpcResponseRegistration<any, any>> = [];
+
+  constructor(private readonly embark: Embark) {
+    this.events = embark.events;
+    this.logger = embark.logger;
+  }
+
+  protected get web3() {
+    return (async () => {
+      if (!this._web3) {
+        await this.events.request2("blockchain:started");
+        // get connection directly to the node
+        const provider = await this.events.request2("blockchain:node:provider", "ethereum");
+        this._web3 = new Web3(provider);
+      }
+      return this._web3;
+    })();
+  }
+
+  private get nodeAccounts() {
+    return (async () => {
+      if (!this._nodeAccounts) {
+        const web3 = await this.web3;
+        this._nodeAccounts = await web3.eth.getAccounts();
+      }
+      return this._nodeAccounts || [];
+    })();
+  }
+
+  private get accounts() {
+    return (async () => {
+      if (!this._accounts) {
+        const web3 = await this.web3;
+        const nodeAccounts = await this.nodeAccounts;
+        this._accounts = AccountParser.parseAccountsConfig(this.embark.config.blockchainConfig.accounts, web3, dappPath(), this.logger, nodeAccounts);
+      }
+      return this._accounts || [];
+    })();
+  }
+
+  public init() {
+    this.registerActions();
+    this.setCommandHandlers();
+  }
+
+  private registerActions() {
+    this.embark.registerActionForEvent("tests:config:updated", { priority: 40 }, (_params, cb) => {
+      // blockchain configs may have changed (ie endpoint)
+      this._web3 = null;
+
+      // web.eth.getAccounts may return a different value now
+      // update accounts across all modifiers
+      this.resetAccounts(cb);
+    });
+
+    this.embark.registerActionForEvent("blockchain:proxy:request", this.onProxyRequest.bind(this));
+    this.embark.registerActionForEvent("blockchain:proxy:response", this.onProxyResponse.bind(this));
+  }
+
+  private setCommandHandlers() {
+    this.events.setCommandHandler("rpc:accounts:reset", this.resetAccounts.bind(this));
+
+    this.events.setCommandHandler("rpc:request:interceptor:register", this.registerRequestInterceptor.bind(this));
+    this.events.setCommandHandler("rpc:response:interceptor:register", this.registerResponseInterceptor.bind(this));
+  }
+
+  private async onProxyRequest<T>(params: ProxyRequestParams<T>, callback: Callback<ProxyRequestParams<T>>) {
+    try {
+      params = await this.executeInterceptors(this.requestInterceptors, params);
+    } catch (err) {
+      err.message = __("Error executing RPC request modifications for '%s': %s", params?.request?.method, err.message);
+      return callback(err);
+    }
+    callback(null, params);
+  }
+
+  private async onProxyResponse<T, R>(params: ProxyResponseParams<T, R>, callback: Callback<ProxyResponseParams<T, R>>) {
```

Comment:
Just a personal preference: when there are two types in a generic method, I prefer to prefix both with `T` (for "Type"), then suffix with a descriptor, ie `TRequest` and `TResponse`. When I read `ProxyResponseParams<TRequest, TResponse>` aloud in my head, I read "Proxy response params of type request and response". However, if it were `ProxyResponseParams<T, R>`, I would read it aloud as "Proxy response params of type T and R", which isn't as immediately obvious to me. Again, just a personal preference.